consensus/parlia/parlia_test.go

```diff
@@ -144,3 +153,449 @@ func randomAddress() common.Address {
 	rand.Read(addrBytes)
 	return common.BytesToAddress(addrBytes)
 }
+
+// =========================================================================
+// =======     Simulator P2P network to verify fast finality    ============
+// =========================================================================
+
+type MockBlock struct {
+	parent *MockBlock
+
+	blockNumber uint64
+	blockHash   common.Hash
+	coinbase    *MockValidator
+	td          uint64 // Total difficulty from genesis block to current block
+	attestation uint64 // Vote attestation for parent block, zero means no attestation
+}
+
+var GenesisBlock = &MockBlock{
+	parent:      nil,
+	blockNumber: 0,
+	blockHash:   common.Hash{},
+	coinbase:    nil,
+	td:          diffInTurn.Uint64(),
+	attestation: 0,
+}
+
+func (b *MockBlock) Hash() (hash common.Hash) {
+	hasher := sha3.NewLegacyKeccak256()
+	rlp.Encode(hasher, []interface{}{
+		b.parent,
+		b.blockNumber,
+		b.coinbase,
+		b.td,
+		b.attestation,
+	})
+	hasher.Sum(hash[:0])
+	return hash
+}
+
+func (b *MockBlock) IsConflicted(a *MockBlock) bool {
+	if a.blockNumber > b.blockNumber {
+		p := a.parent
+		for ; p.blockNumber > b.blockNumber; p = p.parent {
+		}
+
+		return p.blockHash != b.blockHash
+	}
+
+	if a.blockNumber < b.blockNumber {
+		p := b.parent
+		for ; p.blockNumber > a.blockNumber; p = p.parent {
+		}
+
+		return p.blockHash != a.blockHash
+	}
+
+	return a.blockHash != b.blockHash
+}
+
+// GetJustifiedNumberAndHash returns number and hash of the highest justified block,
+// keep same func signature with consensus even if `error` will be nil definitely
+func (b *MockBlock) GetJustifiedNumberAndHash() (uint64, common.Hash, error) {
+	justifiedBlock := GenesisBlock
+	for curBlock := b; curBlock.blockNumber > 1; curBlock = curBlock.parent {
+		// justified
+		if curBlock.attestation != 0 {
+			justifiedBlock = curBlock.parent
+			break
+		}
+
+	}
+
+	return justifiedBlock.blockNumber, justifiedBlock.blockHash, nil
+}
+
+func (b *MockBlock) GetJustifiedNumber() uint64 {
+	justifiedBlockNumber, _, _ := b.GetJustifiedNumberAndHash()
+	return justifiedBlockNumber
+}
+
+// GetFinalizedBlock returns highest finalized block,
+// include current block's attestation.
+func (b *MockBlock) GetFinalizedBlock() *MockBlock {
+	if b.blockNumber < 3 {
+		return GenesisBlock
+	}
+
+	if b.attestation != 0 && b.parent.attestation != 0 {
+		return b.parent.parent
+	}
+
+	return b.parent.GetFinalizedBlock()
+}
+
+type MockValidator struct {
+	index        int
+	validatorSet int // validators number
+	head         *MockBlock
+	voteRecords  map[uint64]*types.VoteData
+}
+
+func NewMockValidator(index int, validatorSet int) *MockValidator {
+	v := &MockValidator{
+		index:        index,
+		validatorSet: validatorSet,
+		head:         GenesisBlock,
+		voteRecords:  make(map[uint64]*types.VoteData),
+	}
+	return v
+}
+
+func (v *MockValidator) SignRecently() bool {
+	parent := v.head
+	for i := 0; i < v.validatorSet*2/3; i++ {
```

Comment:
I believe the testcase here is out of date, can you refresh it?

Reply:
ok

Reply:
these cases have been refreshed after changing new forkchoice
so only this code need to refresh caused by separating chain liveness from finality liveness